### PR TITLE
kvserver: use AdminSplit in splitTestRange

### PR DESCRIPTION
**Pull request by tbg**
`splitTestRange` was previously reaching into the store to finagle a
split. It is used by around a dozen tests. Prototyping around #72374
has shown that these tests frequently need patching up whenever we
adjust (improve) the store's replica handling.

This is a time suck and besides, we also want to be able to test
the Store from within the `kvserver` (not `kvserver_test`) package.
So if we can make that happen, and can use AdminSplit, that would
be preferrable.

AdminSplit requires them to run a (somewhat) distributed multi-range
transaction. A first split would hit a single range, but after that the
split batch hits at least two ranges (meta2 and splitKey), and so we
need nontrivial DistSender-like functionality. Splits are also
nontrivial distributed transactions and so we need a TxnCoordSender.
Experience suggests that it's better to use the "real thing" and to
make sure it's configurable enough to fit the use case, rather than
whipping up half-baked replacements.

Luckily, it turned out that DistSender and TxnCoordSender are already
up to the task, and this commit adopts them in
`createTestStoreWithoutStart`, and changes `splitTestRange` to use
`AdminSplit`.

Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replica_command.go [hunk 1]
```diff
@@ -330,6 +330,7 @@ func (r *Replica) adminSplitWithDescriptor(
 			// correct range.
 			if !kvserverbase.ContainsKey(desc, args.Key) {
 				l, _ := r.GetLease()
+				_ = kvserverbase.ContainsKey(desc, args.Key)
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_command.go:333]: Why this extra call?

**Review by erikgrinaker — APPROVED** [review]

**Review by tbg — COMMENTED** [review]
Wow sorry for posting such a junk PR! It's actually not all junk but I submitted a bunch of stuff that came from my initial attempts at getting splits to work. Removed these extra bits.

bors r=erikgrinaker

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72383#-:-Mnf8wCu4sCN48nZnGCO:bxqiwql)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[pkg/kv/kvserver/replica_command.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72383#-Mna-E8mupvg8eSWTgbQ-r1-333:-Mnf9QTI2KEsnc4IK2PO:b-c01s5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c30b264a2a6f49b7f621e815f9f388c4ec879/pkg/kv/kvserver/replica_command.go#L333)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Why this extra call?
</blockquote></details>

Detritus. At one point I was getting mismatch errors here and I wanted to know why. Since `rr` doesn't work on my platform the easiest way to be able to step into `ContainsKey` with the right args is to duplicate it here. Removed.

**Review by erikgrinaker — COMMENTED** [review]
Yeah, was wondering about the DistSender emulation in the mock senders and stuff, but figured you'd added it temporarily for a reason. Nice to see it gone. :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72383#-:-MnfG81dCPQbHv0v6hY9:bfz1d4m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

**Comment by tbg** [comment]
bors r-

Apparently it was needed for something? Urgh, going to take a look.

**Comment by tbg** [comment]
Oh ok that makes sense. I fixed up the tests that use `createTestStore`, but there are tests using `testContext` that also call `splitTestRange`. So I need to do the thing I wanted to do anyway, which is have `testContext` use the real distributed txn stack as well. Okay, let's see if I can do that.

**Comment by tbg** [comment]
Pretty satisfying - got wrapped up with this just as I'm brushing up against the end of my day. I think this is fit for review though some small thing will certainly be wrong with it. `testContext.StartWithStoreConfigAndVersion` has become a very thin wrapper around `createTestStoreWithoutStart`, which I am quite happy about.

**Review by erikgrinaker — APPROVED** [review]
LGTM at a glance. CI seems green too.

**Comment by tbg** [comment]
tftr!

bors r=erikgrinaker

**Comment by tbg** [comment]
I will stress kvserver before trying again. Pretty sure I've introduced a
few flakes.

(sent from mobile, please excuse my brevity)

**Comment by tbg** [comment]
I went through 200 iters of the kvserver package (thx roachprod-stress) without issues. The one flake I saw during CI is now fixed; it was also flaky on master before.

bors r=erikgrinaker

**Comment by tbg** [comment]
bors r-

ha, overlooked the "21 failures" line of my nonstandard invocation of `make roachprod-stress` 🙈

**Comment by tbg** [comment]
bors r=erikgrinaker
80 runs so far, 0 failures, over 32m20s
I hadn't passed `-p` previously, so I was overloading the machines (probably).

**Pull request merged**